### ISSUE TITLE
xdebug install option

### ISSUE THREAD

**Issue by TimVroom** [ByteInternet/hypernode-vagrant#90]
Hi Byte,

Would it be possible to install xdebug in the vagrant env, e.g just with a option during the local.yml generation?

e.g. Install with xdebug: y/n

Connection can be configured like : 
(at least that is what i internally use), perhaps make the idekey configurable.

Of course a little instruction to configure it on the ide side would be handy, but not required.

```
[XDEBUG]
zend_extension=xdebug.so
xdebug.remote_autostart=0
xdebug.remote_connect_back=1
xdebug.remote_enable=1
xdebug.remote_handler=dbgp
xdebug.remote_port=9000
xdebug.default_enable=0
xdebug.idekey=PHPSTORM
xdebug.profiler_enable=0
xdebug.profiler_enable_trigger=1
xdebug.var_display_max_depth=8
xdebug.collect_vars=1024
```

**Comment by vdloo** [comment]
We've recently implemented blackfire on the production hypernodes [see this knowledgebase post from yesterday](https://support.hypernode.com/knowledgebase/profiling-magento-blackfire/) so I don't think we are going to invest time in installing another profiler right now unfortunately. 

But you are free (or someone else) to implement it yourself if you want to dive in. 

You can define a new option in the interactive vagrant up dialogue by adding it [here](https://github.com/ByteInternet/hypernode-vagrant/blob/master/vagrant/plugins/vagrant-hypconfigmgmt/lib/vagrant-hypconfigmgmt/command.rb#L332) in the same way as the firewall for example. 

You could do something like:
```ruby
    DEFAULT_DEBUG_STATE = false
    AVAILABLE_DEBUG_STATES = [true, false]

    def configure_xdebug(env)
      ensure_setting_exists('xdebug')
      ensure_attribute_configured(
        env, 'xdebug', 'state',
        AVAILABLE_XDEBUG_STATES
      ) { get_xdebug_state(env) }
    end

    def get_xdebug_state(env)
      ask_message = "Do you want to enable xdebug? Enter true or false [default false]: "
      xdebug_enabled = get_setting(env, AVAILABLE_XDEBUG_STATES, DEFAULT_XDEBUG_STATE, ask_message)
      xdebug_state = xdebug_enabled == 'true' ? true : false
      message = "Xdebug will be #{xdebug_state ? 'enabled' : 'disabled'}"
      env[:ui].info(message)
      return xdebug_state
    end
```

and then in the Vagrantfile add a flag to the hypernode.sh [shell provisioner](https://github.com/ByteInternet/hypernode-vagrant/blob/master/Vagrantfile#L60) (or create a new one):
```
    args: "-x #{settings['xdebug']['state']} -m #{settings['magento']['version']} -v #{settings['varnish']['state']} -f #{settings['firewall']['state']}"
```

add a new getopt [here](https://github.com/ByteInternet/hypernode-vagrant/blob/master/vagrant/provisioning/hypernode.sh#L13)

```
        x)
            xdebug_enabled="$OPTARG" ;;
```

and then script the xdebug install [here](https://github.com/ByteInternet/hypernode-vagrant/blob/master/vagrant/provisioning/hypernode.sh#L78)
```
if $xdebug_enabled; then
    # install all the things
fi
```

**Comment by vdloo** [comment]
Hi @TimVroom xdebug has now been added as an option to the local.yml generation prompt, see [this PR](https://github.com/ByteInternet/hypernode-vagrant/pull/113)

```
    hypernode: Do you want to install Xdebug? Enter true or false [default false]: true
==> hypernode: Xdebug will be enabled
```

**Comment by TimVroom** [comment]
Awesome! Really helpfull and seems to be working.

**Comment by tomlankhorst** [comment]
Nice to add this feature. I used apt-get plus some configuration to enable it previously. I'm using a non-default port as port 9000 is unavailable. Guess I still have to do some manual configuration?

**Comment by vdloo** [comment]
Yeah I guess you still have to when using it with an IDE
```
The IDE listens on port 9000, so xdebug.remote_port is set to 9000
```
https://xdebug.org/docs/remote

but we run FPM on 9000
```
app@49bcd4-tmpx2noPI-magweb-vgr:~$ netstat -tulpn | grep 9000
(Not all processes could be identified, non-owned process info
 will not be shown, you would have to be root to see it all.)
tcp        0      0 127.0.0.1:9000          0.0.0.0:*               LISTEN      13086/php-fpm: pool
```

not sure how that's supposed to work, but @fliphess came up with [this configuration](https://support.hypernode.com/knowledgebase/install-xdebug-hypernode-vagrant/). Maybe he can chime in. 

I made this to run the xdebug dependent phpunit tests for the [hypernode-magerun](https://github.com/Hypernode/hypernode-magerun/pull/30/files) plugin so I haven't tested with phpstorm or anything unfortunately.

**Comment by tomlankhorst** [comment]
Well, normally I just set the XDebug port to some alternative number in the IDE (I'm using PhpStorm currently). 

By the way, your IDE might complain about missing path-mappings. Specifically the file `fpm-inspector/prepend.php`. This file is not available outside of the Vagrant box so what I did was copying the file to my PC and configuring a path mapping to it. This solves the issue for me. 

![image](https://cloud.githubusercontent.com/assets/675432/20671894/8c88dd18-b57f-11e6-8bdf-1b235f0f8d59.png)
